### PR TITLE
Allow to add specific shiny-prerendered parts in HTML template

### LINKED ISSUE #2060: Wrong order of dependencies in header for shiny prerendered document ?
This follows #1912 fixed by #1942. 

Following the merge of the PR, @gadenbuie have seen issues in **learnr** possibly related to this: 
* https://github.com/rstudio/learnr/issues/499
* https://github.com/rstudio/learnr/issues/502

Putting his comment here too: https://github.com/rstudio/rmarkdown/pull/1942#discussion_r583672458

> Adding the additional dependencies at the end of `<head>` feels like the right thing to do but, unfortunately, it breaks previous behavior. Before this change the served document structure was (essentially):
> 
> ```
> shiny boilerplate:
> |- head
> |- body
>    |- full prerendered document
>       |- head
>       |- body
> ```
> 
> so the dependencies added by shiny were in the first `<head>` tag and the document's dependencies appeared in the second `<head>` tag. When reconciled by the browser, this ends up being the same as
> 
> ```
> browser DOM
> |- head
>    |- shiny dependencies
>    |- prerendered doc dependencies
> ```
> 
> In learnr, for example, this creates a situation where the scripts and dependencies in the tutorial's pandoc template appear _before_ the tutorial dependencies provided by learnr.
> 
> This may be the correct behavior and we can adjust in learnr, but I suspect there are other templates that take advantage of the implicit ordering of dependencies in the previous behavior.

Currently, format like **learnr** assumes shiny dependencies is loaded before documents dependencies. This is the previous behavior. Maybe we should stick with that. 

This issue is to track  discussion in which order the dependencies should be added. Currently, using shiny templating when the page is render, the dependencies are added at the end of the head 
https://github.com/rstudio/rmarkdown/blob/c0b858496e2346acb82c28db85c0511a0e56dca1/R/shiny_prerendered.R#L163-L169

Should it be at the top of the head ? Or is it more complex than that to correctly handle document head without shiny header ?

cc @schloerke @gadenbuie

### PR DESCRIPTION
**Pull request by cderv**
This should fix #2060 and help solve rstudio/learnr#499 and rstudio/learnr#502 by modifying the **leanr** template.

Instead of adding the placeholder for head content required by `htmltools::renderDocument()` to the end of the `<head>`, we check if it is already in the rendered HTML.

This come with a feature to add content in a Pandoc template conditionnaly to shiny_prerendered rendering. A new pandoc variable `shiny-prerendered` will be set and pass to Pandoc that will allow `$if(shiny-prerendered)$` clause in a Pandoc template. 

All html templates in **rmarkdown** gains addition of `<!-- HEAD_CONTENT -->` placeholder using this. `{{ headContent() }}` is not used because we don't want to pass the all HTML file to `htmltools::htmlTemplate()`. It seems `htmltools::renderDocument()` expect this special comment token anyway, so it seems fine to use it directly. 

With this change, it will be a better fix for #1912 than just #1942 allow greater flexibility for template makers. This will allow **learn** to add the same part in the template and have html dependencies loaded where needed in the resulting html file. 

cc @schloerke @gadenbuie @jcheng5

### CODE REVIEW

File: R/shiny_prerendered.R [hunk 1]
```diff
@@ -160,13 +160,20 @@ shiny_prerendered_html <- function(input_rmd, render_args) {
...
-  # add placeholder for additional dependencies at the end of the <head> element
-  # using a template expansion expected by shiny (usually done in shiny:::renderPage)
-  sub('</head>',
-      paste0('\n', htmltools::htmlTemplate(text_ = "{{ headContent() }}"), '\n</head>'),
+  # The html template used to render the UI should contain the placeholder
+  # expected by shiny in `shiny:::renderPage()` which uses
+  # `htmltools::renderDocument`.
+  # If it is not present in the template, we add this placeholder at the end of
+  # the <head> element
+  if (!any(grepl(headContent <- "<!-- HEAD_CONTENT -->", html_with_deps, fixed = TRUE))) {
```
Comment (schloerke) [inline R/shiny_prerendered.R:168]: Trying to keep the tag isolated to `htmltools`...
```suggestion
  headContent <- as.character(htmltools::htmlTemplate(text_ = "{{ headContent() }}"))
  if (!any(grepl(headContent, html_with_deps, fixed = TRUE))) {
```

**Review by schloerke — APPROVED** [review]
Thank you!

File: R/shiny_prerendered.R (hunk 1 above)
Comment (cderv) [reply R/shiny_prerendered.R:168]: As discussed, we need to use the `<!-- HEAD_CONTENT -->` as raw in the HTML templates to not have them passed to `htmltools::htmlTemplates()`. This means there is no real value to use `headContent()` here as if it changes upstream we need to update the templates. 

I added a test to detect any change on this `headContent()` value in `htmltools::htmlTemplate()`. 

Note: `htmltools::renderDocument()` use the raw comment and documents it as being the token to use with this function. 

Thanks for the feedback!

**Comment by cderv** [comment]
- [ ] still tests to add before merging.

**Comment by cderv** [comment]
@gadenbuie would that work for you too ? 
You would need to add the if close in the **learnr** template and it should fix it.

**Review by gadenbuie — APPROVED** [review]
LGTM, thank you!

**Pull request merged**